core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -189,6 +189,18 @@ private[spark] class TaskSetManager(
     addPendingTask(i)
   }
 
+  {
+    // TaskSet got submitted by DAGScheduler may have some already completed
+    // tasks since DAGScheduler does not always know all the tasks that have
+    // been completed by other tasksets when completing a stage, so we mark
+    // those tasks as finished here to avoid launching duplicate tasks, while
+    // holding the TaskSchedulerImpl lock.
+    // See SPARK-25250 and markPartitionCompletedInAllTaskSets()`
+    sched.stageIdToFinishedPartitions.get(taskSet.stageId).foreach {
+      finishedPartitions => finishedPartitions.foreach(markPartitionCompleted(_, None))
```

Comment:
How do we guarantee that there is no race condition between the creation of `TaskSetManager` and the updating of `TaskSchedulerImpl. stageIdToFinishedPartitions`?

Reply:
we always hold lock on `TaskSchedulerImpl` while accessing `stageIdToFinishedPartitions`:

* `TaskSchedulerImpl.submitTasks`

* `TaskSchedulerImpl.handleSuccessfulTask`

* `TaskSchedulerImpl.taskSetFinished`

Reply:
ah i see, `TaskSetManager` is always created at `* TaskSchedulerImpl.submitTasks`